src/qutip_qip/vqa.py

```diff
@@ -0,0 +1,841 @@
+        return final_state
+
+    def _sample_bitstring_from_state(self, state):
+        """
+        Use probability amplitudes from state after measurement
+        in computational basis to sample a bitstring.
+
+        Parameters
+        ----------
+        state: Qobj
+
+        Returns
+        -------
+        bitstring: str
+            Formatted binary string with length corresponding to
+            dimension of input state
+
+        E.g. the state 1/sqrt(2) * (|0> + |1>)
+        would return 0 and 1 with equal probability.
+        """
+        n_qubits = int(np.log2(state.shape[0]))
+        outcome_indices = list(range(2**n_qubits))
+        probs = [abs(i.item()) ** 2 for i in state]
+        outcome_index = np.random.choice(outcome_indices, p=probs)
+        return format(outcome_index, f"0{n_qubits}b")
+
+    def evaluate_parameters(self, angles):
+        """
+        Evaluate a cost for the circuit, based on the VQA cost
+        method defined.
+
+        Parameters
+        ----------
+        angles: list of float
+            A list of dimension (n,) for n free parameters in the circuit
+
+        Returns
+        -------
+        cost: float
+        """
+        final_state = self.get_final_state(angles)
+        if self.cost_method == "BITSTRING":
+            if self.cost_func is None:
+                raise ValueError("self.cost_func not specified")
+            else:
+                return self.cost_func(
+                    self._sample_bitstring_from_state(final_state)
+                )
+        elif self.cost_method == "STATE":
+            if self.cost_func is None:
+                raise ValueError("self.cost_func not specified")
+            else:
+                return self.cost_func(final_state)
+        elif self.cost_method == "OBSERVABLE":
+            if self.cost_observable is None:
+                raise ValueError("self.cost_observable not specified")
+            else:
+                return expect(self.cost_observable, final_state)
+        else:
+            raise ValueError(f"Unrecognised cost method: {self.cost_method}")
+
+    def optimize_parameters(
+        self,
+        initial="random",
+        method="COBYLA",
+        use_jac=False,
+        do_nothing=False,
+        layer_by_layer=False,
+        bounds=None,
+        constraints=(),
+    ):
+        """
+        Run VQA optimization loop
+
+        Parameters
+        ----------
+        initial: str or list of floats, optional
+            Initialization method for the free parameters.
+            If a list of floats of dimensions (n,) for n free parameters
+            in the circuit is given, then these are taken to be the initial
+            conditions for the optimizer. Otherwise if a string is given:
+            1. "random" will randomize initial free parameters between 0 and 1,
+            2. "ones" will set each initial free parameter to a value of 1.
+            Defaults to "random"
```

Comment:
I've updated this to use bullet points now. Is it correct to space things with newlines like this within the docstring?

```python
"""
[...]
conditions for the optimizer. Otherwise if a string is given:

* (Default) "random" will randomize initial free parameters
  between 0 and 1.

* "ones" will set each initial free parameter to a value of 1.
"""
```

Reply:
Seems to work. 
![image](https://user-images.githubusercontent.com/12125783/164448479-af0ae0cd-b186-4995-af54-52e9d4efab68.png)

You can also download it in our GitHub action 
"Build HTML documentation" from the generated artifacts.(https://github.com/apps/github-actions)@github-actions](https://github.com/qutip/qutip-qip/actions/runs/2201435768)